### PR TITLE
Improvements to the SAR app template

### PR DESCRIPTION
**Pull request by benbridts**
This improves the SAR Application template, I used three commits to make review easier and allow cherry-picking if that would be preferred:

Improve SAR Application template (These are things I discovered by running `cfn-lint -e -cI`):
- replaces every `arn:aws:` with `arn:${AWS::Partition}` (should help with #332 )
- replaces `!Join ['', ...]` with `!Sub`
- added `UpdateReplacePolicy` if there was a `DeletionPolicy`
- changed the type of properties where it didn't match the specification
- removed an unused `!Sub`

Remove invalid permissions: `s3:Get` (without a wildcard) doesn't give any permissions

Use references where possible:
There were still some hardcoded values in the template. I removed those where it was a relatively small change. This should make #342 possible, and make it safer to change names of resources the future.
There are some hardcoded values I didn't change:
- If the resources are created outside of this template
-  `project/aws-deployment-framework-base-templates`, because removing that circle would require converting a managed policy to an inline policy, which is a bigger change

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/template.yml [hunk 1]
```diff
@@ -68,21 +68,22 @@ Resources:
               StringEquals:
                 aws:PrincipalOrgID: !GetAtt Organization.OrganizationId
             Resource:
-              - !Sub arn:aws:s3:::${BootstrapTemplatesBucket}
-              - !Sub arn:aws:s3:::${BootstrapTemplatesBucket}/*
+              - !Ref BootstrapTemplatesBucket
```
Comment (sbkok) [inline src/template.yml:71]: According to the [S3 Policy docs](https://docs.aws.amazon.com/AmazonS3/latest/userguide/s3-arn-format.html), the resource should refer to the Arn of a bucket in S3 Bucket Policies. When you use Ref, according to the [S3 CFn docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-bucket.html#aws-properties-s3-bucket-return-values), it returns the bucket name instead.

I think this line should become:
```yaml
!GetAtt BootstrapTemplatesBucket.Arn
# Or:
!Sub "${BootstrapTemplatesBucket.Arn}"
```

File: src/template.yml [hunk 2]
```diff
@@ -68,21 +68,22 @@ Resources:
...
+              - !Sub "${BootstrapTemplatesBucket.Arn}/*"
             Principal:
               AWS: "*"
           - Action:
               - s3:PutObject*
             Effect: Allow
             Resource:
-              - !Sub arn:aws:s3:::${BootstrapTemplatesBucket}
-              - !Sub arn:aws:s3:::${BootstrapTemplatesBucket}/*
+              - !Ref BootstrapTemplatesBucket
```
Comment (sbkok) [inline src/template.yml:79]: Same here, use `!GetAtt BootstrapTemplatesBucket.Arn` instead.

File: src/template.yml [hunk 3]
```diff
@@ -321,7 +319,6 @@ Resources:
               - "codecommit:List*"
               - "codecommit:CancelUploadArchive"
               - "codecommit:UploadArchive"
-              - "s3:Get"
```
Comment (sbkok) [inline src/template.yml:324]: Yes, that one is not adding any value, thanks for removing.

File: src/template.yml [hunk 4]
```diff
@@ -411,15 +408,17 @@ Resources:
               - "s3:ListBucket"
               - "s3:PutObject"
             Resource:
-              - "arn:aws:s3:::serverlessrepo-aws-deplo-bootstraptemplatesbucket-*"
-              - "arn:aws:s3:::serverlessrepo-aws-deplo-bootstraptemplatesbucket-*/*"
-              - "arn:aws:s3:::adf-shared-modules-*-*"
-              - "arn:aws:s3:::adf-shared-modules-*-*/*"
+              - !Ref BootstrapTemplatesBucket
```
Comment (sbkok) [inline src/template.yml:411]: Same as above, using `!GetAtt BootstrapTemplatesBucket.Arn`.

File: src/template.yml [hunk 5]
```diff
@@ -801,15 +795,27 @@ Resources:
...
+        Version: "2012-10-17"
+        Statement:
+          - Effect: Allow
+            Action: ssm:GetParameter
+            Resource:
+              - !Sub "arn:${AWS::Partition}:ssm:${AWS::Region}:${AWS::AccountId}:parameter/${SharedModulesBucketName}"
+              - !Sub "arn:${AWS::Partition}:ssm:${AWS::Region}:${AWS::AccountId}:parameter/deployment_account_region"
+      Roles:
+        # Created by the Serverless Transform of CrossRegionBucketHandler
+        - !GetAtt CrossRegionBucketHandlerRole
```
Comment (sbkok) [inline src/template.yml:818]: Personally I don't like relying on a to-be-generated transform resource.
This makes reviewing the code very hard to follow.
If this is the only way to break-out of the circular reference, I would rather move the other policies away from the lambda function into a separate role definition as well.

Plus I think this would need to be a `!Ref` instead of a `!GetAtt`.

**Review by sbkok — CHANGES_REQUESTED** [review]
Thanks for cleaning up the template!
I've added a few comments that I think should be improved before we can merge this one.

File: src/template.yml (hunk 5 above)
Comment (benbridts) [reply src/template.yml:818]: I'll keep the hardcoded name for now, making it a dependency can be done in a different PR

File: src/template.yml [hunk 6]
```diff
@@ -434,16 +434,18 @@ Resources:
               - "s3:ListBucket"
               - "s3:PutObject"
             Resource:
-              - !Sub "arn:${AWS::Partition}:s3:::serverlessrepo-aws-deplo-bootstraptemplatesbucket-*"
-              - !Sub "arn:${AWS::Partition}:s3:::serverlessrepo-aws-deplo-bootstraptemplatesbucket-*/*"
+              - !GetAtt "BootstrapTemplatesBucket.Arn"
+              - !Sub "${BootstrapTemplatesBucket.Arn}/*"
```
Comment (benbridts) [inline src/template.yml]: These lines might not have made it in the first version of the PR

**Comment by benbridts** [comment]
@sbkok  I think I fixed everything, I also made a quick overview of the changes that actually reduce permissions:

- L439-440: `arn:aws:s3:::serverlessrepo-aws-deplo-bootstraptemplatesbucket-*` replaced with `BootstrapTemplatesBucket.Arn `
- L443-444: `arn:aws:s3:::adf-shared-modules-*-*` replace with `arn:${AWS::Partition}:s3:::${SharedModulesBucket.BucketName}`
- L833: `arn:aws:s3:::adf-shared-modules-*` replaced with `arn:${AWS::Partition}:s3:::adf-shared-modules-${DeploymentAccountMainRegion}-*`

File: src/template.yml [hunk 7]
```diff
@@ -830,12 +825,15 @@ Resources:
                 - s3:PutEncryptionConfiguration
                 - s3:PutBucketPolicy
                 - s3:PutBucketPublicAccessBlock
-              Resource: "arn:aws:s3:::adf-shared-modules-*"
+              # This must match BucketNamePrefix of the SharedModulesBucket resource
+              Resource: !Sub "arn:${AWS::Partition}:s3:::adf-shared-modules-${DeploymentAccountMainRegion}-*"
```
Comment (sbkok) [inline src/template.yml:829]: When updating, I noticed that the DeploymentAccountMainRegion is not enforced.
If you forget to specify it, it will create a new bucket without the region in its name.
So far, that has not caused any issues. However, when we introduce this change, it will.

So either we should enforce the DeploymentAccountMainRegion in the stack, and make it clear that users should not forget to specify the region. Or we should implement a way to lookup the previous value.
Unfortunately the SAR does not show the values used previously when deploying the same template.

**Review by sbkok — CHANGES_REQUESTED** [review]

**Review by sbkok — APPROVED** [review]
I have updated this PR to reflect the latest changes from our main branch + updated some of the issues that were flagged before but not resolved yet.

It looks good to me to merge. Thanks for contributing.

**Review by StewartW — APPROVED** [review]
Looks great. Really tidies up the template 👍
